### PR TITLE
Improve performance of sum of products with nonlinear expressions

### PR DESCRIPTION
**Pull request by blegat**
Closes https://github.com/jump-dev/JuMP.jl/issues/3469

### CODE REVIEW

**Comment by blegat** [comment]
Looking into an example where it gives an improvement to implement `operate!!` for more than one factor, I found this weird issue:
https://github.com/jump-dev/MutableArithmetics.jl/issues/224

File: src/nlp_expr.jl [hunk 1]
```diff
@@ -368,16 +368,20 @@ for f in (:+, :-, :*, :^, :/, :atan)
...
-    y::AbstractJuMPScalar,
-)
+    args::Vararg{Any,N},
+) where {N}
     _throw_if_not_real(x)
     if x.head == :+
-        push!(x.args, y)
+        arg = *(args...)
+        if _MA.add_sub_op(op) != +
+            arg = _MA.add_sub_op(op)(arg)
```
Comment (odow) [inline src/nlp_expr.jl:379]: I really think we should avoid this. If the user writes `x + y - 2 * z`, we should make `-(+(x, y), *(2, z))`, not `+(x, y, -(*(2, z))). 

Julia has a parsing precedence order. Let's follow that where possible.

**Comment by odow** [comment]
I also don't see how this fixes your issue. It just defines a separate method so that the other issue isn't hit. We'd still need the `if` statement if we added another `AbstractJuMPScalar` that was `IsNotMutable`.

File: src/nlp_expr.jl [hunk 2]
```diff
@@ -368,16 +368,20 @@ for f in (:+, :-, :*, :^, :/, :atan)
 end
 
 function _MA.operate!!(
-    ::typeof(_MA.add_mul),
+    op::_MA.AddSubMul,
     x::GenericNonlinearExpr,
-    y::AbstractJuMPScalar,
```
Comment (odow) [inline src/nlp_expr.jl:373]: I don't remember why I only wrote the binary case 😆

File: src/nlp_expr.jl (hunk 1 above)
Comment (blegat) [reply src/nlp_expr.jl:379]: I thought `flatten!` was doing that anyway

**Pull request closed without merging**